File: 2025/golang/day5/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type IDWindow struct {
	start int
	end   int
}

func addAvailableIds(availableIds []IDWindow, line string) []IDWindow {
	windows := strings.Split(line, "-")
	start, err := strconv.Atoi(windows[0])
	if err != nil {
		panic(err)
	}
	end, err := strconv.Atoi(windows[1])
	if err != nil {
		panic(err)
	}
	return append(availableIds, IDWindow{start: start, end: end})
}

func countAvailableIds(availableIds []IDWindow) int {
	total := 0
	return total
}

func mergeIntervals(availableIds []IDWindow) []IDWindow {
	for i := 0; i < len(availableIds)-1; i++ {
		if availableIds[i].end >= availableIds[i+1].start-1 {
			availableIds[i].end = max(availableIds[i].end, availableIds[i+1].end)
			availableIds = append(availableIds[:i+1], availableIds[i+2:]...)
			i--
		}
	}
	return availableIds
}

func solveDay5Part2(input string) {
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}

	total := 0
	scanner := bufio.NewScanner(file)
	availableIds := []IDWindow{}
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) == 0 {
			break
		}
		availableIds = addAvailableIds(availableIds, line)
	}
	sort.Slice(availableIds, func(i, j int) bool {
		return availableIds[i].start < availableIds[j].start
	})

	availableIds = mergeIntervals(availableIds)
	for _, id := range availableIds {
		total += id.end - id.start + 1
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	fmt.Println("Total: ", total)
}

func main() {
	// input := "input_test.txt" // Expected output: 14
	input := "input.txt" // 357485433193284
	solveDay5Part2(input)
}

File: 2025/cpp/day5/part2.cpp
#include <iostream>
#include <fstream>
#include <vector>



typedef struct {
    long long start;
    long long end;
} IDWindow;


int is_fresh(std::vector<IDWindow> &available_ids, long long line) {
    for (auto &available_id: available_ids) {
        if (line >= available_id.start && line <= available_id.end) {
            return 1;
        }
    }
    return 0;
}

void merge_intervals(std::vector<IDWindow> &available_ids) {
    using namespace std;

    sort(available_ids.begin(), available_ids.end(), [](const IDWindow &a, const IDWindow &b) {
        return a.start < b.start;
    });
    for (size_t i = 0; i < -1; i++) {
        if (i >= available_ids.size() - 1) {
            break;
        }
        if (available_ids[i].end >= available_ids[i+1].start-1) {
            available_ids[i].end = max(available_ids[i].end, available_ids[i+1].end);
            available_ids.erase(available_ids.begin() + i + 1);
            i--;
        }
    }
}

void add_available_ids(std::vector<IDWindow> &available_ids, std::string line) {
    long long start;
    long long end;
    size_t del_position = line.find('-');
    start = std::stoll(line.substr(0, del_position));
    end = std::stoll(line.substr(del_position + 1));

    available_ids.push_back(IDWindow{start, end});
}

void solve_day5_part1(std::string input_file) {
    using namespace std;

    long long total = 0;
    vector<IDWindow> available_ids;
    ifstream input(input_file);
    string line;

    while (getline(input, line)) {
        if (line.empty()) {
            break;
        }
        add_available_ids(available_ids, line);
    }

    merge_intervals(available_ids);

    for (auto &available_id: available_ids) {
        total += (available_id.end - available_id.start + 1);
    }

    cout << "Total: " << total << endl;
}

int main() {
    // std::string input_file = "input_test.txt"; // Expected output: 14
    std::string input_file = "input.txt"; // 357485433193284
    solve_day5_part1(input_file);

    return 0;
}
